chain/postgresdb/database/database.go

```diff
@@ -0,0 +1,334 @@
+			case uniqueViolation:
+				return ErrDBDuplicatedEntry
+			}
+		}
+		return err
+	}
+
+	return nil
+}
+
+// QuerySlice is a helper function for executing queries that return a
+// collection of data to be unmarshalled into a slice.
+func QuerySlice[T any](ctx context.Context, l log.Logger, db sqlx.ExtContext, query string, dest *[]T) error {
+	return NamedQuerySlice(ctx, l, db, query, struct{}{}, dest)
+}
+
+// NamedQuerySlice is a helper function for executing queries that return a
+// collection of data to be unmarshalled into a slice where field replacement is necessary.
+func NamedQuerySlice[T any](ctx context.Context, l log.Logger, db sqlx.ExtContext, query string, data any, dest *[]T) error {
+	q := queryString(query, data)
+
+	if _, ok := data.(struct{}); ok {
+		//nolint:gomnd // Having to add constants is overkill.
+		l.AddCallerSkip(2).Infow("database.QuerySlice", "query", q)
+	} else {
+		l.AddCallerSkip(1).Infow("database.QuerySlice", "query", q)
+	}
+
+	rows, err := sqlx.NamedQueryContext(ctx, db, query, data)
+	if err != nil {
+		//nolint:errorlint // We want this
+		if pqerr, ok := err.(*pq.Error); ok && pqerr.Code == undefinedTable {
+			return ErrUndefinedTable
+		}
+		return err
+	}
+	defer rows.Close()
+
+	var slice []T
+	for rows.Next() {
+		v := new(T)
+		if err := rows.StructScan(v); err != nil {
+			return err
+		}
+		slice = append(slice, *v)
+	}
+	*dest = slice
+
+	return nil
+}
+
+// QueryStruct is a helper function for executing queries that return a
+// single value to be unmarshalled into a struct type where field replacement is necessary.
+func QueryStruct(ctx context.Context, l log.Logger, db sqlx.ExtContext, query string, dest any) error {
+	return NamedQueryStruct(ctx, l, db, query, struct{}{}, dest)
+}
+
+// NamedQueryStruct is a helper function for executing queries that return a
+// single value to be unmarshalled into a struct type where field replacement is necessary.
+func NamedQueryStruct(ctx context.Context, l log.Logger, db sqlx.ExtContext, query string, data, dest any) error {
+	q := queryString(query, data)
+
+	if _, ok := data.(struct{}); ok {
+		//nolint:gomnd // This doesn't need to be a constant
+		l.AddCallerSkip(2).Infow("database.QueryStruct", "query", q)
+	} else {
+		l.AddCallerSkip(1).Infow("database.QueryStruct", "query", q)
+	}
+
+	rows, err := sqlx.NamedQueryContext(ctx, db, query, data)
+	if err != nil {
+		//nolint:errorlint // Have no clue why this is a problem :(.
+		if pqerr, ok := err.(*pq.Error); ok && pqerr.Code == undefinedTable {
+			return ErrUndefinedTable
+		}
+		return err
+	}
+	defer rows.Close()
+
+	if !rows.Next() {
+		return ErrDBNotFound
+	}
+
+	if err := rows.StructScan(dest); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// queryString provides a pretty print version of the query and parameters.
+func queryString(query string, args ...any) string {
+	query, params, err := sqlx.Named(query, args)
+	if err != nil {
+		return err.Error()
```

Comment:
Will this result in an attempt at the error being run as a query? If so, that might result in some confusing error message

Reply:
Great catch!

Reply:
This function is for logging and not executing the query. The idea is to log that error.